Include/internal/pycore_bitutils.h

```diff
@@ -82,6 +84,31 @@ _Py_bswap64(uint64_t word)
 }
 
 
+// Population count: count the number of 1's in 'u'
+// (number of bits set to 1)
+static inline int
+_Py_popcount32(uint32_t u)
+{
+#if defined(__clang__) || defined(__GNUC__)
+
+#if SIZEOF_INT == 4
+    return __builtin_popcount(u);
+#elif SIZEOF_LONG == 4
+    return __builtin_popcountl(u);
+#else
+#  error "unsupported configuration"
+#endif
+
+#else
+    /* 32-bit SWAR (SIMD Within A Register) popcount */
+    u -= (u >> 1) & UINT32_C(0x55555555);
+    u = (u & UINT32_C(0x33333333)) + ((u >> 2) & UINT32_C(0x33333333));
+    u = (u + (u >> 4)) & UINT32_C(0x0f0f0f0f);
+    return (uint32_t)(u * UINT32_C(0x01010101)) >> 24;
```

Comment:
My original version is actually safer here. :-) This version can still invoke undefined behaviour (albeit only on unusual machines). Realistically, it's unlikely that we'd ever hit the undefined behaviour in practice, but given that there's an easy way to write this that avoids that undefined behaviour, I'd prefer to use that way.

If your `int` type has width larger than 32 bits, then C's integer promotions (C99 §6.3.1.1) ensure that both 
`u` and `UINT32_C(0x01010101)` are treated as (signed) `int` *before* the multiply, and then there's potential for the multiply to overflow, which gives undefined behaviour.

In contrast, if in the same situation you do `u * 0x01010101U`, then after those same integer promotions you're multiplying an `int` by an `unsigned int`, and C's "usual arithmetic conversions" (C99 §6.3.1.8) kick in to ensure that the multiplication is actually performed as `unsigned int` by `unsigned int`, which is safe from undefined behaviour and simply discards high bits in the expected manner if necessary.

Performance isn't likely to be a problem with the `(uint32_t)(u * 0x01010101U)` version: the cast ensures that a compiler doesn't have to be all that clever to realise that it's enough to do a 32x32->32 unsigned multiply here.

And yes, it's crazy that it's *this* hard to write a portable uint32_t * uint32_t -> uint32_t multiply in standard C. Maybe it's time to rewrite Python's core in Rust.

Reply:
One more safe solution is:
```
return (uint32_t)((uint64_t)u * (uint64_t)0x01010101)
```
again relying on the compiler to recognise that it's not necessary to perform a full 64-bit by 64-bit multiplication.

Some of those casts are redundant, of course: we only need *one* of the inner casts (but the code is easier to reason about if both are present) and if the return type is already declared as `uint32_t` then we don't need the outer cast, but again it helps with reading and reasoning.

Reply:
I used (uint32_t)(uint64_t * uint64_t).